File: goblobBase/src/test/java/com/cauchymop/goblob/presenter/GamePresenterTest.java
package com.cauchymop.goblob.presenter;

import com.cauchymop.goblob.model.Analytics;
import com.cauchymop.goblob.model.GameDatas;
import com.cauchymop.goblob.model.GameRepository;
import com.cauchymop.goblob.model.GoGameController;
import com.cauchymop.goblob.proto.PlayGameData;
import com.cauchymop.goblob.view.GameView;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static com.cauchymop.goblob.model.TestDataHelperKt.createGameData;
import static com.cauchymop.goblob.proto.PlayGameData.GameData.Phase.INITIAL;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;

@RunWith(MockitoJUnitRunner.class)
public class GamePresenterTest {
  private static final GameDatas GAME_DATAS = new GameDatas();

  @Mock
  private Analytics analytics;
  @Mock
  private GameRepository gameRepository;
  @Mock
  private AchievementManager achievementManager;
  @Mock
  private GameView view;
  @Mock
  private GameViewUpdater gameViewUpdater;
  @Mock
  private ConfigurationViewEventProcessor configurationViewEventProcessor;
  @Mock
  private InGameViewEventProcessor inGameViewEventProcessor;
  @Mock
  private GoGameController goGameController;
  @Mock
  private FeedbackSender feedbackSender;
  @Mock
  private SingleGamePresenter singleGamePresenter;

  private GamePresenter gamePresenter;

  @Before
  public void setUp() throws Exception {
    gamePresenter = createGamePresenter();
    gamePresenter.setView(view);
    reset(view, gameViewUpdater, gameRepository);
  }

  @After
  public void tearDown() throws Exception {
    verifyNoMoreInteractions(analytics, gameRepository, achievementManager, gameViewUpdater, view, goGameController);
  }

  @Test
  public void setView_registersAsGameRepositoryListener() {
    gamePresenter.setView(view);

    verify(gameViewUpdater).setView(view);
    verify(gameRepository).addGameSelectionListener(gamePresenter);
  }


  @Test
  public void gameSelected_updatesView() throws Exception {
    PlayGameData.GameData gameData = createGameData().setMatchId("pipo").setPhase(INITIAL).build();

    gamePresenter.gameSelected(gameData);

    verify(gameViewUpdater).setGoGameController(any());
    verify(view).setConfigurationViewListener(any());
    verify(view).setInGameActionListener(any());
    verify(goGameController).setGameData(gameData);

    verify(gameRepository).addGameChangeListener(any());
    verify(achievementManager).updateAchievements(goGameController);
    verify(gameViewUpdater).update();
  }

  @Test
  public void gameSelected_withNullGameData_noPreviousGame_doesNothing() throws Exception {

    gamePresenter.gameSelected(null);

  }

  @Test
  public void gameSelected_withNullGameData_andPreviousGame_clearsPreviousGame() throws Exception {
    setInitialGame();

    gamePresenter.gameSelected(null);

    // From SingleGamePresenter.clear()
    verify(gameRepository).removeGameChangeListener(any());

  }

  @Test
  public void gameSelected_withPreviousGame_clearPreviousGame_andUpdatesView() throws Exception {
    setInitialGame();
    PlayGameData.GameData selectedGameData = createGameData().setMatchId("pipo").setPhase(INITIAL).build();

    gamePresenter.gameSelected(selectedGameData);

    verify(gameViewUpdater).setGoGameController(any());
    verify(view).setConfigurationViewListener(any());
    verify(view).setInGameActionListener(any());
    verify(goGameController).setGameData(selectedGameData);

    verify(gameRepository).addGameChangeListener(any());
    verify(achievementManager).updateAchievements(goGameController);
    verify(gameViewUpdater).update();
    // From SingleGamePresenter.clear()
    verify(gameRepository).removeGameChangeListener(any());
  }

  @Test
  public void clear_withNoGame() throws Exception {
    gamePresenter.clear();

    verify(gameRepository).removeGameSelectionListener(gamePresenter);
    verify(view).setConfigurationViewListener(null);
    verify(view).setInGameActionListener(null);
  }

  @Test
  public void clear_withGame() throws Exception {
    setInitialGame();

    gamePresenter.clear();

    verify(gameRepository).removeGameSelectionListener(gamePresenter);
    verify(view).setConfigurationViewListener(null);
    verify(view).setInGameActionListener(null);
    // From SingleGamePresenter.clear()
    verify(gameRepository).removeGameChangeListener(any());
  }

  public void setInitialGame() {
    gamePresenter.gameSelected(createGameData().build());
    reset(gameRepository, achievementManager, gameViewUpdater, view, goGameController);
  }

  private GamePresenter createGamePresenter() {
    return new GamePresenter(GAME_DATAS,
        analytics,
        gameRepository,
        gameViewUpdater,
        feedbackSender,
        goGameController,
        singleGamePresenter);
  }
}

File: goblobBase/src/test/java/com/cauchymop/goblob/presenter/GamePresenterTest.kt
package com.cauchymop.goblob.presenter

import com.cauchymop.goblob.model.*
import com.cauchymop.goblob.proto.PlayGameData.GameData.Phase.INITIAL
import com.cauchymop.goblob.view.GameView
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.BDDMockito
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class GamePresenterTest {

  @Mock
  private lateinit var analytics: Analytics

  @Mock
  private lateinit var gameRepository: GameRepository

  @Mock
  private lateinit var achievementManager: AchievementManager

  @Mock
  private lateinit var view: GameView

  @Mock
  private lateinit var configurationViewEventProcessor: ConfigurationViewEventProcessor

  @Mock
  private lateinit var inGameViewEventProcessor: InGameViewEventProcessor

  @Mock
  private lateinit var goGameController: GoGameController

  @Mock
  private lateinit var feedbackSender: FeedbackSender

  @Mock
  private lateinit var gameViewUpdater: GameViewUpdater


  private lateinit var gamePresenter: GamePresenter

  @Before
  fun setUp() {
    gamePresenter = createGamePresenter()
    gamePresenter.view = view
    reset<Any>(view, gameViewUpdater, gameRepository, achievementManager, goGameController)
  }

  @After
  fun tearDown() {
    verify(goGameController, atLeast(0)).matchId
    verifyNoMoreInteractions(analytics, gameRepository, achievementManager, gameViewUpdater, view, goGameController)
  }

  @Test
  fun setView() {
    gamePresenter.view = view

    verify(gameViewUpdater).view = view
    verify(gameRepository).addGameSelectionListener(gamePresenter)
    verify(gameRepository).addGameChangeListener(gamePresenter)
    verify(view).setConfigurationViewListener(configurationViewEventProcessor)
    verify(view).setInGameActionListener(inGameViewEventProcessor)
    // from gameUpdated
    verify(achievementManager).updateAchievements(goGameController)
    verify(gameViewUpdater).update()
  }


  @Test
  fun gameSelected_withGameData() {
    val gameData = createGameData().setMatchId("pipo").setPhase(INITIAL).build()

    gamePresenter.gameSelected(gameData)

    verify(goGameController).gameData = gameData
    verify(gameViewUpdater).update()
  }

  @Test
  fun gameSelected_withNoGameData_doesNothing() {

    gamePresenter.gameSelected(null)

  }


  @Test
  fun gameChanged_withDifferentMatchId_doesNothing() {
    setInitialGame("pizza")

    gamePresenter.gameChanged(createGameData().setMatchId("pipo").build())

    // Does nothing.
  }

  @Test
  fun clear() {
    gamePresenter.clear()

    verify(gameRepository).removeGameChangeListener(gamePresenter)
    verify(gameRepository).removeGameSelectionListener(gamePresenter)
    verify(view).setConfigurationViewListener(null)
    verify(view).setInGameActionListener(null)
  }

  @Test
  fun gameChanged_withSameMatchId() {
    setInitialGame("pizza")

    gamePresenter.gameChanged(createGameData().setMatchId("pizza").setPhase(INITIAL).build())

    verify(gameViewUpdater).update()
    verify(achievementManager).updateAchievements(goGameController)
  }

  private fun setInitialGame(matchId: String) {
    BDDMockito.given(goGameController.matchId).willReturn(matchId)
  }

  fun setInitialGame() {
    gamePresenter.gameSelected(createGameData().build())
    reset<Any>(gameRepository, achievementManager, gameViewUpdater, view, goGameController)
  }

  private fun createGamePresenter(): GamePresenter {
    return GamePresenter(gameRepository,
        achievementManager,
        gameViewUpdater,
        configurationViewEventProcessor,
        inGameViewEventProcessor,
        goGameController)
  }

  companion object {
    private val GAME_DATAS = GameDatas()
  }
}
